File: python/rocketmq/signature.py
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from rocketmq.client_config import ClientConfig
import uuid
import datetime
import importlib.metadata
from rocketmq.utils import sign


class Signature:
    __AUTHORIZATION_KEY = "authorization"

    __DATE_TIME_KEY = "x-mq-date-time"
    __DATE_TIME_FORMAT = "%Y%m%dT%H%M%SZ"

    __SESSION_TOKEN_KEY = "x-mq-session-token"
    __CLIENT_ID_KEY = "x-mq-client-id"
    __REQUEST_ID_KEY = "x-mq-request-id"
    __LANGUAGE_KEY = "x-mq-language"
    __CLIENT_VERSION_KEY = "x-mq-client-version"
    __PROTOCOL_VERSION = "x-mq-protocol"

    __ALGORITHM = "MQv2-HMAC-SHA1"
    __CREDENTIAL = "Credential"
    __SIGNED_HEADERS = "SignedHeaders"
    __SIGNATURE = "Signature"

    @staticmethod
    def sign(client_config: ClientConfig, client_id: str):
        date_time = datetime.datetime.now().strftime(Signature.__DATE_TIME_FORMAT)
        metadata = [
            (Signature.__LANGUAGE_KEY, "PYTHON"),
            (Signature.__PROTOCOL_VERSION, "v2"),
            (Signature.__CLIENT_VERSION_KEY, importlib.metadata.version("rocketmq")),
            (
                Signature.__DATE_TIME_KEY,
                date_time,
            ),
            (Signature.__REQUEST_ID_KEY, uuid.uuid4()),
            (Signature.__CLIENT_ID_KEY, client_id),
        ]
        if not client_config.session_credentials_provider:
            return metadata
        session_credentials = (
            client_config.session_credentials_provider.session_credentials()
        )
        if not session_credentials:
            return metadata
        if session_credentials.security_token:
            metadata.append(
                (Signature.__SESSION_TOKEN_KEY, session_credentials.security_token)
            )
        if (not session_credentials.access_key) or (
            not session_credentials.access_secret
        ):
            return metadata
        signature = sign(session_credentials.access_key, date_time)
        authorization = (
            Signature.__ALGORITHM
            + " "
            + Signature.__CREDENTIAL
            + "="
            + session_credentials.access_key
            + ", "
            + Signature.__SIGNED_HEADERS
            + "="
            + Signature.__DATE_TIME_KEY
            + ", "
            + Signature.__SIGNATURE
            + "="
            + signature
        )
        metadata.append((Signature.__AUTHORIZATION_KEY, authorization))
        return metadata

File: php/Signature.php
<?php
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace Apache\Rocketmq;

/**
 * Signature - Generates gRPC metadata with MQv2-HMAC-SHA1 authentication headers.
 * Mirrors Java Signature.sign(ClientConfiguration config, ClientId clientId).
 *
 * Algorithm:
 * 1. Build baseline metadata (language, protocol, version, datetime, request-id, etc.)
 * 2. Optionally add STS security token
 * 3. Compute HMAC-SHA1(accessSecret, dateTime) -> lowercase hex digest
 * 4. Build authorization header:
 *    "MQv2-HMAC-SHA1 Credential=<AK>, SignedHeaders=x-mq-date-time, Signature=<hex>"
 */
class Signature
{
    private const ALGORITHM = 'MQv2-HMAC-SHA1';
    private const CREDENTIAL = 'Credential';
    private const SIGNED_HEADERS = 'SignedHeaders';
    private const SIGNATURE = 'Signature';
    private const DATE_TIME_FORMAT = 'Ymd\THis\Z';
    private const SIGNED_HEADERS_VALUE = 'x-mq-date-time';

    /**
     * Generate signed gRPC metadata.
     *
     * @param SessionCredentials|null $credentials
     * @param string $clientId
     * @param string $language Language string (e.g., "PHP")
     * @param string $clientVersion Client version string (e.g., "5.0.0")
     * @param string $namespace Namespace string
     * @param string $protocol Protocol version (e.g., "v2")
     * @return array gRPC metadata array
     */
    public static function sign(
        ?SessionCredentials $credentials,
        string $clientId,
        string $language = 'PHP',
        string $clientVersion = '5.0.0',
        string $namespace = '',
        string $protocol = 'v2'
    ): array {
        $dateTime = gmdate(self::DATE_TIME_FORMAT);
        $requestId = self::generateUUID();

        $metadata = [
            'x-mq-client-id' => [$clientId],
            'x-mq-language' => [$language],
            'x-mq-client-version' => [$clientVersion],
            'x-mq-protocol' => [$protocol],
            'x-mq-date-time' => [$dateTime],
            'x-mq-request-id' => [$requestId],
            'x-mq-namespace' => [$namespace],
        ];

        if ($credentials !== null) {
            // Add STS security token if present
            $securityToken = $credentials->getSecurityToken();
            if (!empty($securityToken)) {
                $metadata['x-mq-session-token'] = [$securityToken];
            }

            // Compute HMAC-SHA1 signature
            $accessSecret = $credentials->getAccessSecret();
            $accessKey = $credentials->getAccessKey();
            $signature = self::hmacSha1($accessSecret, $dateTime);

            // Build authorization header value
            $authValue = self::ALGORITHM
                . ' ' . self::CREDENTIAL . '=' . $accessKey
                . ', ' . self::SIGNED_HEADERS . '=' . self::SIGNED_HEADERS_VALUE
                . ', ' . self::SIGNATURE . '=' . $signature;

            $metadata['authorization'] = [$authValue];
        }

        return $metadata;
    }

    /**
     * Compute HMAC-SHA1 and return lowercase hex digest.
     * Mirrors Java TLSHelper.sign(accessSecret, dateTime).
     */
    private static function hmacSha1(string $key, string $data): string
    {
        $raw = hash_hmac('sha1', $data, $key, true);
        return self::encodeHex($raw);
    }

    /**
     * Encode binary data as lowercase hex string.
     */
    private static function encodeHex(string $data): string
    {
        $hex = '';
        $len = strlen($data);
        for ($i = 0; $i < $len; $i++) {
            $byte = ord($data[$i]);
            $hex .= ($byte < 16 ? '0' : '') . dechex($byte);
        }
        return $hex;
    }

    /**
     * Generate a UUID v4 string for request-id.
     */
    private static function generateUUID(): string
    {
        return sprintf(
            '%08x-%04x-%04x-%04x-%012x',
            mt_rand(0, 0xffffffff),
            mt_rand(0, 0xffff),
            mt_rand(0, 0xffff) & 0x0fff | 0x4000,
            mt_rand(0, 0x3fff) | 0x8000,
            mt_rand(0, 0xffffffffffff)
        );
    }
}
